errors/multi.go

```diff
@@ -0,0 +1,145 @@
+package errors
+
+import (
+	"fmt"
+	"strings"
+)
+
+var _ coder = (*multiErr)(nil)
+var _ causer = (*multiErr)(nil)
+var _ Multi = (*multiErr)(nil)
+
+// Multi is an interface for multiErr to avoid exposing actual implementation
+// to the rest of the app.
+type Multi interface {
+	// Add adds an error to this multiErr
+	Add(err error)
+	// AddNamed adds an error that could later be retrieved by name
+	AddNamed(name string, err error)
+	// Named returns a named error or nil
+	Named(name string) error
+	// IsEmpty returns true if there are no actual errors registered,
+	// this would also return true when registering a bunch of nil errors
+	IsEmpty() bool
+	error
+}
+
+// multiErr is a default implementation of errors.Multi.
+// It does not support flattening in order to maintain consistent
+// behaviour of named errors. The implementation is not thread-safe
+type multiErr struct {
+	errors     []error
+	errorNames map[string]int
+}
+
+func (me *multiErr) IsEmpty() bool {
+	return len(me.errors) == 0
+}
+
+func (me *multiErr) first() error {
+	if me.IsEmpty() {
+		return nil
+	}
+	return me.errors[0]
+}
+
+func (me *multiErr) Add(err error) {
+	if err == nil {
+		return
+	}
+	me.errors = append(me.errors, err)
+}
+
+// This AddNamed implementation would overwrite a pointer to
+// a named error if the same name is used twice, while keeping
+// the original error in the container for matching.
+func (me *multiErr) AddNamed(name string, err error) {
+	if err == nil {
+		return
+	}
+	me.errors = append(me.errors, err)
+	me.errorNames[name] = len(me.errors) - 1
+}
+
+func (me *multiErr) Named(name string) error {
+	index, ok := me.errorNames[name]
+	if ok {
+		return me.errors[index]
+	}
+	return nil
+}
+
+func (me *multiErr) Error() string {
+	if me.IsEmpty() {
+		return ""
+	}
+
+	if len(me.errors) == 1 {
+		return me.first().Error()
+	}
+
+	errs := make([]string, len(me.errors))
+	for i, err := range me.errors {
+		errs[i] = fmt.Sprintf("* %s", err)
+	}
+
+	return fmt.Sprintf(
+		"%d errors occurred:\n\t%s\n\n",
+		len(me.errors), strings.Join(errs, "\n\t"))
+}
+
+// is provides a helper for Error.Is to work with multiErr
+func (me *multiErr) is(isFunc func(error) bool) bool {
+	if me.IsEmpty() {
+		return isFunc(nil)
+	}
+
+	res := false
+	for _, err := range me.errors {
+		res = isFunc(err)
+		if res {
+			return res
+		}
+	}
+	return res
+}
+
+// ABCICode returns the error code of a first error consistent with fail-fast approach or falls back to
+// internalError code if the error does not satisfy the coder interface.
+// Returns success code if the multiErr is empty
+func (me *multiErr) ABCICode() uint32 {
+	return abciCode(me.first())
```

Comment:
This is for me the biggest issue. I do not think that returning a random error code is good enough. This approach returns a bad error code. Client will be confused if an error with the description "invalid currency" returns _ErrUnauthorized_ code or "user name is required" with _ErrCurrency_. This approach does not solve the original issue of returning a random from the user perspective error.

This is why I have insisted on having a single code for validation errors. I think it is much better to use a single code like _ErrValidation_ with a set of human readable descirptions to signal that the validation failed. It will be obvious for the client what happened - validation failed.

Reply:
I disagree. I see no point in having ambiguous validation error code as it's not actionable. 

I see how this decision is both easy and compelling, but I see no good use-case for this. Also, this basically converts half of the errors we throw into one errorCode.

The approach that is currently taken is an assumption that had to be made as there seems to be no better alternative. If there's a better proposal that does not involve hardcoding an ABCICode for all the multi/validation errors - I'm all ears.

Consider this:
1. We have a bullet list of all the errors, first of which would correspond to the code, if this is not straightforward enough - we can point that out in the description better
2. Error descriptions aka logs are only being used by people debugging the client or otherwise using it programatically, codes, however, could be actionable on some level by the consumer
3. We return code that is consistent with current fail-fast approach, where we get the code of a first validation error, so for the client nothing really changes, except they get follow-up error descriptions too, in case there's a human that actually reads that. Once the first error is resolved - we get the second one etc, which keeps being consistent with current approach.
4. It's not a random error, it's the first one. Stack trace would be a good analogy for this. It's useful.
5. This feature has been designed mostly for simplifying the test code as far as I remember, which it does to a certain extent. I don't think the idea was to obfuscate error codes with some generic one.

Reply:
I agree with Piotr on the abciCode. The client needs to be able to distinguish between a single  Invalid currency and a multiErr with additional errors. This has direct impact on how the error(s) can be displayed to the users in an UI.
Therefore I would recommend a dedicated error code for MultiErr and a structured serialization format of the content(s) for the ABCIResponse.Log so that any client can do the following:
```
if abci.code == mulitiErrCode{
  for i, v := range deserialize(abci.log){ // flattens nested multierrs if we support this but better not
    fmt.Printf("#%d : Code %d, msg: %s", i, v.code, v.msg)
  }
}else{
   fmt.Printf(Code %d, msg: %s",  abci.code, abci.log)
}
```